### PR TITLE
*: add a PERFORMANCE_SCHEMA.slow_query table

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

Create a virtual table `TiNiuB.slow_query`

```
mysql> show databases;
+--------------------+
| Database           |
+--------------------+
| INFORMATION_SCHEMA |
| PERFORMANCE_SCHEMA |
| TiNiuB             |
| mysql              |
| test               |
+--------------------+
5 rows in set (0.01 sec)
```

```
mysql> use tiniub;
Database changed
mysql> select * from slow_query order by duration limit 2;
+-----------------+----------------------------+----------+---------+------+---------+----------------+----------------+--------+-----------+-----------+----------+
| SQL             | START                      | DURATION | DETAILS | SUCC | CONN_ID | TRANSACTION_TS | USER           | DB     | TABLE_IDS | INDEX_IDS | INTERNAL |
+-----------------+----------------------------+----------+---------+------+---------+----------------+----------------+--------+-----------+-----------+----------+
| select sleep(1) | 2019-01-08 16:14:26.540384 | 00:00:01 |         |    1 |       1 |              0 | root@127.0.0.1 | tiniub |           |           |        0 |
| select sleep(2) | 2019-01-08 16:14:31.577712 | 00:00:02 |         |    1 |       1 |              0 | root@127.0.0.1 | tiniub |           |           |        0 |
+-----------------+----------------------------+----------+---------+------+---------+----------------+----------------+--------+-----------+-----------+----------+
2 rows in set (0.00 sec)
```

Compare to the `admin show slow` method, this virtual table if more flexible.

```
mysql> admin show slow recent 2;
+-----------------+----------------------------+----------+---------+------+---------+----------------+----------------+--------+-----------+-----------+----------+
| SQL             | START                      | DURATION | DETAILS | SUCC | CONN_ID | TRANSACTION_TS | USER           | DB     | TABLE_IDS | INDEX_IDS | INTERNAL |
+-----------------+----------------------------+----------+---------+------+---------+----------------+----------------+--------+-----------+-----------+----------+
| select sleep(3) | 2019-01-08 16:14:35.718031 | 00:00:03 |         |    1 |       1 |              0 | root@127.0.0.1 | tiniub |           |           |        1 |
| select sleep(2) | 2019-01-08 16:14:31.577712 | 00:00:02 |         |    1 |       1 |              0 | root@127.0.0.1 | tiniub |           |           |        1 |
+-----------------+----------------------------+----------+---------+------+---------+----------------+----------------+--------+-----------+-----------+----------+
2 rows in set (0.00 sec)
```

Borrow ideas from https://github.com/TiNiuB/tiquery
This could serve as a plugin demo later @lysu 

### What is changed and how it works?

`slow_query` table is similar to `perfschema` tables, except that it's not standard, so a plugin maybe more applicable.

Note that `tiniub.Init()` is not called in this PR, so the `TiNiuB` database is not created by default.

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

### CODE REVIEW

**Comment by morgo** [comment]
May I suggest changing the duration to `ms`?  On some systems 100ms is a slow query.

File: infoschema/tiniub/init.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+		if err != nil {
+			panic(err)
+		}
+		meta.ID = autoid.GenLocalSchemaID()
+		for _, c := range meta.Columns {
+			c.ID = autoid.GenLocalSchemaID()
+		}
+		dbInfo := &model.DBInfo{
+			ID:      autoid.GenLocalSchemaID(),
+			Name:    model.NewCIStr("TiNiuB"),
```
Comment (jackysp) [inline infoschema/tiniub/init.go]: Will the server fail to start if there is already a database which name is also 'TiNiuB'?
Maybe we should give some suggestions to the user when this happened: 1. disable this plugin and start again, 2. rename the existed database.
btw: it is better to use an English name, not a Chinese Pinyin as the database name.

File: infoschema/infoschema.go [hunk 2]
```diff
@@ -358,5 +358,13 @@ func initInfoSchemaDB() {
 
 // IsMemoryDB checks if the db is in memory.
 func IsMemoryDB(dbName string) bool {
-	return dbName == "information_schema" || dbName == "performance_schema"
+	if dbName == "information_schema" {
+		return true
+	}
+	for _, driver := range drivers {
```
Comment (lysu) [inline infoschema/infoschema.go]: when do we call `tiniub.Init()`? and do we have race-condition to access this `drivers`? for current plugin is start time is ok...

but if drivers can be dynamic change it's better to take care race.

File: infoschema/tiniub/init.go (hunk 1 above)
Comment (lysu) [reply infoschema/tiniub/init.go]: re, it's better give `tiniub` a English name? or can we direct put new table into `PERFORMANCE_SCHEMA` or `INFOMATION_SCHEMA`?

In mysql, we can write a `information_schema plugin` to  add a table to information_schema, https://dev.mysql.com/doc/refman/5.7/en/writing-information-schema-plugins.html

File: infoschema/tiniub/tables.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+		meta: meta,
+		cols: columns,
+	}
+	return t
+}
+
+// IterRecords rewrites the IterRecords method of slowQueryTable.
+func (s slowQueryTable) IterRecords(ctx sessionctx.Context, startKey kv.Key, cols []*table.Column, fn table.RecordIterFunc) error {
+	dom := domain.GetDomain(ctx)
+	result := dom.ShowSlowQuery(&ast.ShowSlow{Tp: 42})
```
Comment (lysu) [inline infoschema/tiniub/tables.go]: `42` is a great number maybe added in `ShowSlowType`? :laughing:

File: infoschema/tiniub/tables.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+		row = append(row, types.NewDurationDatum(types.Duration{item.Duration, types.MaxFsp}))
+		row = append(row, types.NewDatum(item.Detail.String()))
+		row = append(row, types.NewDatum(item.Succ))
+		row = append(row, types.NewDatum(item.ConnID))
+		row = append(row, types.NewDatum(item.TxnTS))
+		row = append(row, types.NewDatum(item.User))
+		row = append(row, types.NewDatum(item.DB))
+		row = append(row, types.NewDatum(item.TableIDs))
+		row = append(row, types.NewDatum(item.IndexIDs))
+		row = append(row, types.NewDatum(item.Internal))
```
Comment (lysu) [inline infoschema/tiniub/tables.go]: I found it's maybe useful to combine https://github.com/pingcap/parser/pull/32's sql digest into here, then we can agg a group similar sql's time.

**Comment by winoros** [comment]
I just wonder that whether this pr will be merged into master?
This means that one plugin may need to change TiDB's own code?

**Comment by lysu** [comment]
@tiancaiamao 

I had tried to move tables part to https://github.com/lysu/slow_log_table

and integration it in https://github.com/pingcap/tidb/pull/9006 (https://github.com/lysu/tidb/commit/a8b1f3f0ed51121ed795a4758743b3c7fee75d9b)

It seems work at first glance - - but needs more refining.

**Comment by morgo** [comment]
I have some feedback with naming (sorry, not all of it is easily actionable):
* The `performance_schema` name for queries is "statements".  Since it is possible for future aggregations of this (by user, by account, by host) I think we should call it something like `statements_slow_global`.
* The column `user` should be called `account` to match pfs terminology.
* Using MySQL guidelines, this would technically not be a pfs table but a `sys` table.  pfs is the raw structured event data (although it can have an aggregation).  If you filter it or order it for a DBA perspective, that is `sys`.  So for example in MySQL 5.7, there the `statement_analysis` view:
```
mysql> select * from statement_analysis LIMIT 1\G
*************************** 1. row ***************************
            query: SELECT `cat` . `name` AS `CATA ... s_database` ( `sch` . `name` )
               db: performance_schema
        full_scan: *
       exec_count: 1
        err_count: 0
       warn_count: 0
    total_latency: 6.53 ms
      max_latency: 6.53 ms
      avg_latency: 6.53 ms
     lock_latency: 6.10 ms
        rows_sent: 5
    rows_sent_avg: 5
    rows_examined: 26
rows_examined_avg: 26
    rows_affected: 0
rows_affected_avg: 0
       tmp_tables: 1
  tmp_disk_tables: 0
      rows_sorted: 5
sort_merge_passes: 0
           digest: d76cf3ff2efdc088db44d6c7e099dd54911416168582241c1aea8a2188e05e66
       first_seen: 2019-01-10 08:12:28.923857
        last_seen: 2019-01-10 08:12:28.923857
1 row in set (0.00 sec)
```

**Comment by lysu** [comment]
LGTM

**Comment by lysu** [comment]
@tiancaiamao maybe need remove tiniub/tables.go and it will be package as plugin

**Comment by tiancaiamao** [comment]
Hi @morgo @winoros 
We'll make a plugin first, and if the plugin works well, it can serve as canonical plugin example.
Due to the time constraints (plugin feature demo on the devcon), we can fulfill @lysu 's work later https://github.com/lysu/slow_log_table/.

**Pull request closed without merging**